### PR TITLE
AVRO-2763:Resource leak:'<unassigned Closeable value>' is never closed

### PR DESCRIPTION
**Pull request by zeshuai007**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO/) issues and references them in the PR title. For example, "AVRO-1234: My Avro PR"
  - https://issues.apache.org/jira/browse/AVRO-2763
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: doc/examples/java-example/src/main/java/example/GenericMain.java [hunk 1]
```diff
@@ -66,6 +66,7 @@ public static void main(String[] args) throws IOException {
       user = dataFileReader.next(user);
       System.out.println(user);
     }
+    dataFileReader.close();
```
Comment (Fokko) [inline doc/examples/java-example/src/main/java/example/GenericMain.java]: Can we use a try-with-resource pattern here?

File: doc/examples/java-example/src/main/java/example/SpecificMain.java [hunk 2]
```diff
@@ -68,6 +68,7 @@ public static void main(String[] args) throws IOException {
       user = dataFileReader.next(user);
       System.out.println(user);
     }
+    dataFileReader.close();
```
Comment (Fokko) [inline doc/examples/java-example/src/main/java/example/SpecificMain.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/TestDataFile.java [hunk 3]
```diff
@@ -247,15 +247,17 @@ private void testReadWithHeader() throws IOException {
     reader = DataFileReader.openReader(sin, new GenericDatumReader<>(), header, false);
     assertEquals("Should not move from sync point on reopen", validPos, sin.tell());
     assertNotNull("Should be able to reopen at sync point", reader.next());
+    reader.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/TestDataFile.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/TestDataFileCorruption.java [hunk 4]
```diff
@@ -90,5 +90,6 @@ public void testCorruptedFile() throws IOException {
     assertEquals("endive", r.next().toString());
     assertEquals("fig", r.next().toString());
     assertFalse(r.hasNext());
+    r.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/TestDataFileCorruption.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/TestDataFileDeflate.java [hunk 5]
```diff
@@ -59,5 +59,6 @@ public void testWriteAndRead() throws IOException {
     assertEquals("bye bye world", r.next().toString());
     assertEquals("bye bye moon", r.next().toString());
     assertFalse(r.hasNext());
+    r.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/TestDataFileDeflate.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/TestDataFileMeta.java [hunk 6]
```diff
@@ -58,13 +59,15 @@ public void testUseMeta() throws IOException {
     assertTrue(r.getMetaKeys().contains("hello"));
 
     assertEquals("bar", r.getMetaString("hello"));
+    r.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/TestDataFileMeta.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/TestDataFileMeta.java [hunk 7]
```diff
@@ -58,13 +59,15 @@ public void testUseMeta() throws IOException {
...
   }
 
   @Test(expected = AvroRuntimeException.class)
   public void testUseMetaAfterCreate() throws IOException {
     DataFileWriter<?> w = new DataFileWriter<>(new GenericDatumWriter<>());
     w.create(Schema.create(Type.NULL), new ByteArrayOutputStream());
     w.setMeta("foo", "bar");
+    w.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/TestDataFileMeta.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/TestSchemaBuilder.java [hunk 8]
```diff
@@ -792,6 +792,7 @@ public void testDefaults() throws IOException {
     Assert.assertEquals(13, rec2read.get("nullableIntWithDefault"));
     Assert.assertNull(rec2read.get("newOptionalInt"));
     Assert.assertEquals(5, rec2read.get("newNullableIntWithDefault"));
+    reader.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/TestSchemaBuilder.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestNonStringMapKeys.java [hunk 9]
```diff
@@ -254,6 +254,7 @@ public void testRecordNameInvariance() throws Exception {
         fail("Fail with schema: " + schema);
       }
     }
+    fileReader.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestNonStringMapKeys.java]: Can we use a try-with-resource pattern here?

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestNonStringMapKeys.java [hunk 10]
```diff
@@ -274,6 +275,7 @@ public void testRecordNameInvariance() throws Exception {
     while (fileReader.hasNext()) {
       records.add(fileReader.next(record));
     }
+    fileReader.close();
```
Comment (Fokko) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestNonStringMapKeys.java]: Can we use a try-with-resource pattern here?

File: lang/java/ipc-netty/src/test/java/org/apache/avro/ipc/netty/TestNettyServer.java [hunk 11]
```diff
@@ -176,6 +176,7 @@ public void testBadRequest() throws IOException {
     byte[] buf = new byte[2048];
     int bytesRead = sock.getInputStream().read(buf);
     Assert.assertTrue("Connection should have been closed", bytesRead == -1);
+    sock.close();
```
Comment (Fokko) [inline lang/java/ipc-netty/src/test/java/org/apache/avro/ipc/netty/TestNettyServer.java]: Can we use a try-with-resource pattern here?

File: lang/java/ipc/src/test/java/org/apache/avro/TestProtocolSpecific.java [hunk 12]
```diff
@@ -342,6 +342,7 @@ public void testClient() throws Exception {
         proto.testEchoBytes();
         proto.testError();
         System.out.println("Done! Validation java client to " + f.getName() + " - " + port);
+        reader.close();
```
Comment (Fokko) [inline lang/java/ipc/src/test/java/org/apache/avro/TestProtocolSpecific.java]: Can we use a try-with-resource pattern here?

File: lang/java/mapred/src/test/java/org/apache/avro/hadoop/io/TestAvroSequenceFile.java [hunk 13]
```diff
@@ -121,6 +122,7 @@ public void testReadAvroWithoutReaderSchemas() throws IOException {
     assertEquals(2, value.datum().intValue());
 
     assertNull("Should be no more records.", reader.next(key));
+    reader.close();
```
Comment (Fokko) [inline lang/java/mapred/src/test/java/org/apache/avro/hadoop/io/TestAvroSequenceFile.java]: Can we use a try-with-resource pattern here?

File: lang/java/mapred/src/test/java/org/apache/avro/hadoop/io/TestAvroSequenceFile.java [hunk 14]
```diff
@@ -81,6 +81,7 @@ public void testReadAvro() throws IOException {
...
```
Comment (Fokko) [inline lang/java/mapred/src/test/java/org/apache/avro/hadoop/io/TestAvroSequenceFile.java]: Can we use a try-with-resource pattern here?

**Comment by zeshuai007** [comment]
OK, I have changed all to try-with-resource. Thanks for the review @Fokko !

**Comment by anhldbk** [comment]
@zeshuai007 @Fokko  Hi guys. A naive question: Would you please tell me how to properly open a Pull Request for Apache Avro? I see not so much discussion on your issue: https://issues.apache.org/jira/browse/AVRO-2763. Is there any guideline?

**Comment by Fokko** [comment]
No worries @anhldbk. Here's a guide: https://opensource.com/article/19/7/create-pull-request-github Please let me know if there are any further questions. You have to fork the repository, and then you can create a PR on your own fork, from which you can create a PR.

**Comment by anhldbk** [comment]
@Fokko Still I think we need discussions with maintainers first via Jira issues. Would you please have a look at [my opening issue](https://issues.apache.org/jira/browse/AVRO-2723?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17031652#). Thank you in advance.

**Comment by anhldbk** [comment]
@Fokko Travis have failed. I checked the logs and I think we have problems with service quality of Travis

**Comment by Fokko** [comment]
@anhldbk can you rebase onto master?

**Comment by zeshuai007** [comment]
@Fokko
Sure, it has been rebase.

**Pull request merged**